File: aimaster-desktop/services/python-audio/app/profiling/validate.py
"""
Runtime validator for ReferenceProfile v1 dicts.

Mirrors packages/shared-types/src/profile/validate.ts.  Enforces the
"no identifying metadata, no reconstruction-enabling data" invariants
at deserialisation time.
"""
from __future__ import annotations

import re
from typing import Any

from .schema import REFERENCE_PROFILE_SCHEMA_ID

_SLUG_RE = re.compile(r"^[A-Za-z0-9_-]{2,128}$")
_SEMVER_RE = re.compile(r"^\d+\.\d+\.\d+(?:-[A-Za-z0-9.-]+)?$")
_HEX64_RE = re.compile(r"^[a-f0-9]{64}$")

_FORBIDDEN_PROVENANCE_KEYS = {"artist", "title", "album", "lyrics", "isrc", "mbid"}


class ProfileValidationError(ValueError):
    def __init__(self, errors: list[dict[str, str]]) -> None:
        self.errors = errors
        joined = "; ".join(f"{e['path']}: {e['message']}" for e in errors)
        super().__init__(f"ReferenceProfile validation failed — {joined}")


def _err(errs: list[dict[str, str]], path: str, msg: str) -> None:
    errs.append({"path": path, "message": msg})


def _num(v: Any) -> bool:
    return isinstance(v, (int, float)) and not isinstance(v, bool)


def _percentiles(errs: list[dict[str, str]], v: Any, path: str) -> None:
    if not isinstance(v, dict):
        _err(errs, path, "must be object with p10/p50/p90")
        return
    for k in ("p10", "p50", "p90"):
        if not _num(v.get(k)):
            _err(errs, f"{path}/{k}", "must be number")


def validate_reference_profile(data: Any) -> list[dict[str, str]]:
    """Return list of error dicts.  Empty ⇒ valid."""
    errs: list[dict[str, str]] = []
    if not isinstance(data, dict):
        return [{"path": "", "message": "profile must be a JSON object"}]

    if data.get("$schema") != REFERENCE_PROFILE_SCHEMA_ID:
        _err(errs, "/$schema", f"must equal {REFERENCE_PROFILE_SCHEMA_ID}")

    if not isinstance(data.get("id"), str) or not _SLUG_RE.match(data["id"]):
        _err(errs, "/id", "must be slug")

    if not isinstance(data.get("version"), str) or not _SEMVER_RE.match(data["version"]):
        _err(errs, "/version", "must be semver")

    # Provenance — enforce the no-identifying-metadata rule.
    prov = data.get("provenance")
    if not isinstance(prov, dict):
        _err(errs, "/provenance", "object required")
    else:
        for forb in _FORBIDDEN_PROVENANCE_KEYS:
            if forb in prov:
                _err(
                    errs,
                    f"/provenance/{forb}",
                    "forbidden identifying field (schema requires no rights-holder metadata)",
                )
        if prov.get("sourceType") not in (
            "user-supplied", "fixture", "builtin", "streaming-snapshot",
        ):
            _err(errs, "/provenance/sourceType", f"unknown sourceType: {prov.get('sourceType')!r}")
        if not isinstance(prov.get("createdAt"), str):
            _err(errs, "/provenance/createdAt", "must be ISO string")
        if not _num(prov.get("durationSec")) or prov["durationSec"] <= 0:
            _err(errs, "/provenance/durationSec", "must be positive number")
        if not _num(prov.get("sampleRate")) or prov["sampleRate"] <= 0:
            _err(errs, "/provenance/sampleRate", "must be positive number")
        if not _num(prov.get("channels")) or not (1 <= prov["channels"] <= 8):
            _err(errs, "/provenance/channels", "must be int 1..8")
        if not isinstance(prov.get("extractorVersion"), str):
            _err(errs, "/provenance/extractorVersion", "must be string")
        sha = prov.get("sourceFileSha256")
        if sha is not None and not (isinstance(sha, str) and _HEX64_RE.match(sha)):
            _err(errs, "/provenance/sourceFileSha256", "must be null or 64-char hex sha256")

    # Features — enforce the no-fingerprint, no-time-series rule.
    f = data.get("features")
    if not isinstance(f, dict):
        _err(errs, "/features", "object required")
        return errs

    # Loudness
    ld = f.get("loudness")
    if not isinstance(ld, dict):
        _err(errs, "/features/loudness", "object required")
    else:
        for k in ("integratedLufs", "truePeakDbtp", "loudnessRange"):
            if not _num(ld.get(k)):
                _err(errs, f"/features/loudness/{k}", "must be number")
        _percentiles(errs, ld.get("shortTermPercentiles"), "/features/loudness/shortTermPercentiles")
        _percentiles(errs, ld.get("momentaryPercentiles"), "/features/loudness/momentaryPercentiles")

    # Dynamics
    dy = f.get("dynamics")
    if not isinstance(dy, dict):
        _err(errs, "/features/dynamics", "object required")
    else:
        for k in ("crestDb", "transientDensityPerMin"):
            if not _num(dy.get(k)):
                _err(errs, f"/features/dynamics/{k}", "must be number")
        cs = dy.get("compressionScore")
        if not _num(cs) or cs < 0 or cs > 1:
            _err(errs, "/features/dynamics/compressionScore", "must be number in [0, 1]")

    # Tonal — bin count cap
    tn = f.get("tonal")
    if not isinstance(tn, dict):
        _err(errs, "/features/tonal", "object required")
    else:
        spec = tn.get("thirdOctSpectrumDb")
        if not isinstance(spec, dict):
            _err(errs, "/features/tonal/thirdOctSpectrumDb", "object required")
        else:
            if len(spec) > 64:
                _err(
                    errs,
                    "/features/tonal/thirdOctSpectrumDb",
                    f"bin count {len(spec)} > 64 — 1/3-oct resolution maximum (no fingerprinting)",
                )
            for k, v in spec.items():
                if not _num(v):
                    _err(errs, f"/features/tonal/thirdOctSpectrumDb/{k}", "must be number")
        for k in ("spectralTiltDbPerOct", "lowMidBalanceDb", "vocalRegionEnergyDb",
                  "airBandEnergyDb", "harshnessIndex"):
            if not _num(tn.get(k)):
                _err(errs, f"/features/tonal/{k}", "must be number")
        sub = tn.get("subEnergyRatio")
        if not _num(sub) or sub < 0 or sub > 1:
            _err(errs, "/features/tonal/subEnergyRatio", "must be in [0, 1]")

    # Stereo
    st = f.get("stereo")
    if not isinstance(st, dict):
        _err(errs, "/features/stereo", "object required")
    else:
        for k in ("msRatioDb", "stereoWidthIndex"):
            if not _num(st.get(k)):
                _err(errs, f"/features/stereo/{k}", "must be number")
        c = st.get("correlationMean")
        if not _num(c) or c < -1.001 or c > 1.001:
            _err(errs, "/features/stereo/correlationMean", "must be in [-1, +1]")

    # featureFingerprint
    fp = data.get("featureFingerprint")
    if not isinstance(fp, str) or not _HEX64_RE.match(fp):
        _err(errs, "/featureFingerprint", "must be 64-char hex sha256")

    # Guard: arrays in features (forbidden — could be time-series or fingerprint).
    for k, v in (f or {}).items():
        if isinstance(v, list) and len(v) > 200:
            _err(
                errs,
                f"/features/{k}",
                f"array length {len(v)} > 200 — forbidden (no time-series / no fingerprint arrays)",
            )

    return errs

File: aimaster-desktop/packages/shared-types/src/profile/validate.ts
// Runtime validator for ReferenceProfile v1.  No external deps.

import {
  REFERENCE_PROFILE_SCHEMA_ID,
  type ReferenceProfile,
} from './profile.js';

export interface ProfileValidationError {
  path: string;
  message: string;
}

export interface ProfileValidationResult {
  ok: boolean;
  errors: ProfileValidationError[];
  profile?: ReferenceProfile;
}

const isObj = (v: unknown): v is Record<string, unknown> =>
  typeof v === 'object' && v !== null && !Array.isArray(v);
const isStr = (v: unknown): v is string => typeof v === 'string';
const isNum = (v: unknown): v is number =>
  typeof v === 'number' && Number.isFinite(v);

const SLUG_RE = /^[A-Za-z0-9_-]{2,128}$/;
const SEMVER_RE = /^\d+\.\d+\.\d+(?:-[A-Za-z0-9.-]+)?$/;

function err(errs: ProfileValidationError[], path: string, msg: string): void {
  errs.push({ path, message: msg });
}

function validatePercentiles(
  errs: ProfileValidationError[],
  v: unknown,
  path: string,
): void {
  if (!isObj(v)) {
    err(errs, path, 'must be object with p10/p50/p90');
    return;
  }
  for (const k of ['p10', 'p50', 'p90'] as const) {
    if (!isNum(v[k])) err(errs, `${path}/${k}`, 'must be number');
  }
}

/**
 * SAFETY GUARD — reject profiles that smell like fingerprints / contain
 * forbidden identifying fields.  Implementations MUST keep these out of
 * shared profiles; this checker enforces the schema's "no-identification"
 * promise at deserialisation time.
 */
function validateNoForbiddenFields(
  errs: ProfileValidationError[],
  raw: Record<string, unknown>,
): void {
  const root = raw;
  const provenance = (isObj(root['provenance']) ? root['provenance'] : {}) as Record<string, unknown>;
  for (const forbidden of ['artist', 'title', 'album', 'lyrics', 'isrc', 'mbid']) {
    if (forbidden in provenance) {
      err(
        errs,
        `/provenance/${forbidden}`,
        `forbidden identifying field (must not appear in ReferenceProfile)`,
      );
    }
  }
  // Detect things that look like fingerprint arrays (large byte sequences).
  const f = root['features'];
  if (isObj(f)) {
    for (const k of Object.keys(f)) {
      const v = (f as Record<string, unknown>)[k];
      if (Array.isArray(v) && v.length > 200) {
        err(
          errs,
          `/features/${k}`,
          `array length ${v.length} > 200 — this schema forbids long arrays (no time-series, no fingerprint)`,
        );
      }
    }
  }
}

export function validateReferenceProfile(input: unknown): ProfileValidationResult {
  const errs: ProfileValidationError[] = [];
  if (!isObj(input)) {
    return { ok: false, errors: [{ path: '', message: 'profile must be an object' }] };
  }

  if (input['$schema'] !== REFERENCE_PROFILE_SCHEMA_ID) {
    err(errs, '/$schema', `must equal ${REFERENCE_PROFILE_SCHEMA_ID}`);
  }
  if (!isStr(input['id']) || !SLUG_RE.test(input['id'])) {
    err(errs, '/id', 'must be a slug');
  }
  if (!isStr(input['version']) || !SEMVER_RE.test(input['version'])) {
    err(errs, '/version', 'must be semver');
  }

  validateNoForbiddenFields(errs, input);

  const prov = input['provenance'];
  if (!isObj(prov)) {
    err(errs, '/provenance', 'object required');
  } else {
    if (!['user-supplied', 'fixture', 'builtin', 'streaming-snapshot'].includes(
      prov['sourceType'] as string,
    )) {
      err(errs, '/provenance/sourceType', `unknown sourceType: ${prov['sourceType']}`);
    }
    if (!isStr(prov['createdAt'])) err(errs, '/provenance/createdAt', 'must be ISO string');
    if (!isNum(prov['durationSec']) || (prov['durationSec'] as number) <= 0) {
      err(errs, '/provenance/durationSec', 'must be positive number');
    }
    if (!isNum(prov['sampleRate'])) err(errs, '/provenance/sampleRate', 'must be number');
    if (!isNum(prov['channels'])) err(errs, '/provenance/channels', 'must be number');
    if (!isStr(prov['extractorVersion'])) err(errs, '/provenance/extractorVersion', 'must be string');
    const sha = prov['sourceFileSha256'];
    if (sha !== null && sha !== undefined) {
      if (!isStr(sha) || (sha as string).length !== 64) {
        err(errs, '/provenance/sourceFileSha256', 'must be null or 64-char hex sha256');
      }
    }
  }

  const f = input['features'];
  if (!isObj(f)) {
    err(errs, '/features', 'object required');
  } else {
    // Loudness
    const ld = f['loudness'];
    if (!isObj(ld)) {
      err(errs, '/features/loudness', 'object required');
    } else {
      for (const k of ['integratedLufs', 'truePeakDbtp', 'loudnessRange'] as const) {
        if (!isNum(ld[k])) err(errs, `/features/loudness/${k}`, 'must be number');
      }
      validatePercentiles(errs, ld['shortTermPercentiles'], '/features/loudness/shortTermPercentiles');
      validatePercentiles(errs, ld['momentaryPercentiles'], '/features/loudness/momentaryPercentiles');
    }

    // Dynamics
    const dy = f['dynamics'];
    if (!isObj(dy)) {
      err(errs, '/features/dynamics', 'object required');
    } else {
      for (const k of ['crestDb', 'transientDensityPerMin'] as const) {
        if (!isNum(dy[k])) err(errs, `/features/dynamics/${k}`, 'must be number');
      }
      const cs = dy['compressionScore'];
      if (!isNum(cs) || cs < 0 || cs > 1) {
        err(errs, '/features/dynamics/compressionScore', 'must be number in [0, 1]');
      }
    }

    // Tonal
    const tn = f['tonal'];
    if (!isObj(tn)) {
      err(errs, '/features/tonal', 'object required');
    } else {
      if (!isObj(tn['thirdOctSpectrumDb'])) {
        err(errs, '/features/tonal/thirdOctSpectrumDb', 'object required (bin -> dB)');
      } else {
        const bins = Object.keys(tn['thirdOctSpectrumDb'] as Record<string, unknown>);
        if (bins.length > 64) {
          err(
            errs,
            '/features/tonal/thirdOctSpectrumDb',
            `bin count ${bins.length} > 64 — 1/3-oct resolution maximum (no fingerprinting)`,
          );
        }
      }
      for (const k of ['spectralTiltDbPerOct', 'subEnergyRatio', 'lowMidBalanceDb', 'vocalRegionEnergyDb', 'airBandEnergyDb', 'harshnessIndex'] as const) {
        if (!isNum(tn[k])) err(errs, `/features/tonal/${k}`, 'must be number');
      }
      const sub = tn['subEnergyRatio'];
      if (isNum(sub) && (sub < 0 || sub > 1)) {
        err(errs, '/features/tonal/subEnergyRatio', 'must be in [0, 1]');
      }
    }

    // Stereo
    const st = f['stereo'];
    if (!isObj(st)) {
      err(errs, '/features/stereo', 'object required');
    } else {
      for (const k of ['correlationMean', 'msRatioDb', 'stereoWidthIndex'] as const) {
        if (!isNum(st[k])) err(errs, `/features/stereo/${k}`, 'must be number');
      }
      const c = st['correlationMean'];
      if (isNum(c) && (c < -1.001 || c > 1.001)) {
        err(errs, '/features/stereo/correlationMean', 'must be in [-1, +1]');
      }
    }
  }

  if (!isStr(input['featureFingerprint']) || (input['featureFingerprint'] as string).length !== 64) {
    err(errs, '/featureFingerprint', 'must be 64-char hex sha256');
  }

  if (errs.length > 0) return { ok: false, errors: errs };
  return { ok: true, errors: [], profile: input as unknown as ReferenceProfile };
}
